src/main/scala/com/johnsnowlabs/nlp/annotators/spell/context/ContextSpellCheckerApproach.scala

```diff
@@ -0,0 +1,427 @@
+    var maxWid = 0
+    for(word <-sorted) {
+      if(acc < currBinLimit){
+        acc += word._2
+        classes = classes.updated(word._1, (currClass, currWordId))
+        currWordId += 1
+      }
+      else{
+        acc += word._2
+        currClass += 1
+        currBinLimit = (currClass + 1) * binMass
+        classes = classes.updated(word._1, (currClass, 0))
+        currWordId = 1
+      }
+      if (currWordId > maxWid)
+        maxWid = currWordId
+    }
+
+    require(maxWid  < k, "The language model is unbalanced, pick a larger" +
+        s" number of classes using setLMClasses(), current value is ${getOrDefault(languageModelClasses)}.")
+
+    logger.info(s"Max num of words per class: $maxWid")
+    classes
+  }
+
+  /*
+  *  here we do some pre-processing of the training data, and generate the vocabulary
+  * */
+
+  def genVocab(dataset: Dataset[_]) = {
+
+    import dataset.sparkSession.implicits._
+    // for every sentence we have one end and one begining
+    val eosBosCount = dataset.count()
+
+    var vocab = collection.mutable.HashMap(dataset.select(getInputCols.head).as[Seq[Annotation]].
+      flatMap(identity).
+      groupByKey(_.result).
+      mapGroups{case (token, insts) => (token, insts.length.toDouble)}.
+      collect(): _*)
+
+    // words appearing less that minCount times will be unknown
+    val unknownCount = vocab.filter(_._2 < getOrDefault(minCount)).values.sum
+
+    // remove unknown tokens
+    vocab = vocab.filter(_._2 >= getOrDefault(minCount))
+
+    // second pass: identify tokens that belong to special classes, and replace with a label
+    vocab.foreach { case (word, count) =>
+      getOrDefault(specialClasses).foreach { specialClass =>
+        // check word is in vocabulary and the word is uncommon
+        if(specialClass.inVocabulary(word)){
+          if(count < getOrDefault(classCount)) {
+            logger.debug(s"Recognized $word as ${specialClass.label}")
+            vocab.get(word).map { count => vocab.update(specialClass.label, count + 1.0)
+            }.getOrElse(vocab.update(specialClass.label, 1.0))
+
+            // remove the token from global vocabulary, now it's covered by the class
+            vocab.remove(word)
+          } else {
+            specialClass match {
+              // remove the word from the class, it's already covered by the global vocabulary
+              case p:VocabParser => p.vocab.remove(word)
+              case _ =>
+            }
+          }
+       }
+      }
+    }
+
+    /* Blacklists {fwis, hyphen, slash} */
+    // words that appear with first letter capitalized (e.g., at the beginning of sentence)
+    val fwis = vocab.filter(_._1.length > 1).filter(_._1.head.isUpper).
+      filter(w => vocab.contains(w._1.head.toLower + w._1.tail)).keys
+
+    // Remove compound words for which components parts are in vocabulary, keep only the high frequent ones
+    val compoundWords = Seq("-", "/").flatMap { separator =>
```

Comment:
can this list be a param?

Reply:
If this turns out to be a new param, let's do it in RC3 after I release RC2